File: modules/rust_workloads/benches/bench_quicksort.rs
//! Quicksort bench: recursive call tree renders as a `quicksort → partition`
//! flamegraph tower, with a branch-miss story on random input.

use std::hint::black_box;

use criterion::{BatchSize, Criterion, criterion_group, criterion_main};
use pprof::criterion::{Output, PProfProfiler};
use rust_workloads::{quicksort, workload_n};

const DEFAULT_N: usize = 1_000_000;

fn bench_quicksort(c: &mut Criterion) {
    let n = workload_n(DEFAULT_N);
    let input = quicksort::random_vec(n, 42);
    c.bench_function("quicksort", |bench| {
        bench.iter_batched_ref(
            || input.clone(),
            |v| quicksort::quicksort(black_box(v)),
            BatchSize::LargeInput,
        );
    });
}

criterion_group! {
    name = benches;
    // 10 samples: one iteration sorts 1M elements, so the default 100 samples
    // cannot fit criterion's 5s measurement window.
    config = Criterion::default()
        .sample_size(10)
        .with_profiler(PProfProfiler::new(100, Output::Protobuf));
    targets = bench_quicksort
}
criterion_main!(benches);

File: modules/java_workloads/benches/BenchQuicksort.java
// Quicksort bench: recursive call tree renders as a quicksort →
// partition flamegraph tower, with a branch-miss story on random input.

package monorepo.workloads;

import java.util.Arrays;
import java.util.concurrent.TimeUnit;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@State(Scope.Benchmark)
@Fork(1)
public class BenchQuicksort {
    private static final int DEFAULT_N = 1_000_000;

    private long[] input;

    @Setup
    public void setup() {
        input = Quicksort.randomSlice(WorkloadN.workloadN(DEFAULT_N), 42);
    }

    @Benchmark
    public long[] quicksort() {
        long[] buf = Arrays.copyOf(input, input.length);
        Quicksort.quicksort(buf);
        return buf;
    }
}
